### PR TITLE
Immutable config: alter Delivery interface

### PR DESCRIPTION
**Pull request by fractalwrench**
## Goal

We wish to make the `Delivery` interface more resilient to changes in the number of parameters, and to better signify whether a request was successful or not.

## Changeset

Removed the `config.errorApiHeaders` and `config.sessionApiHeaders` that a user would have to call manually, and replaced them with a `DeliveryParams` class that contains the endpoint to post to, and the headers for the request.

Additionally, the `deliver` method now returns a `DeliveryStatus` enum that signifies whether a request has succeeded, failed, or needs retrying. This replaces the previous more implicit API which relied on a specific exception being thrown.

Overall, a custom implementation of the method now looks like this:

```
override fun deliver(report: Report, deliveryParams: DeliveryParams): DeliveryStatus {
    deliveryParams.endpoint // URL to post to
    deliveryParams.headers // headers to include in request
    return DeliveryStatus.DELIVERED // return enum signifying whether delivery succeed/failed/needs retrying
}
```

Note: this is a *breaking change* to the interface.

## Tests

Existing mazerunner scenarios and tests were updated to use the new delivery mechanism.

### CODE REVIEW

File: bugsnag-android-core/src/main/java/com/bugsnag/android/DefaultDelivery.kt [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            return DeliveryStatus.FAILURE
+        } finally {
+            IOUtils.close(conn)
+        }
+    }
+
+    internal fun getDeliveryStatus(responseCode: Int): DeliveryStatus {
+        return when (responseCode) {
+            in 200..299 -> DeliveryStatus.DELIVERED
+            in 400..499 -> DeliveryStatus.FAILURE
```
Comment (bengourley) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/DefaultDelivery.kt]: Although I don't think our api will ever respond with these status codes, for correctness `429` and `408` can (and should) be retried

https://github.com/bugsnag/bugsnag-js/blob/5e241be93dc03a7bd536aa9ecaa4f588738358a3/packages/delivery-expo/delivery.js#L131-L140

File: bugsnag-android-core/src/main/java/com/bugsnag/android/Configuration.java [hunk 2]
```diff
@@ -729,13 +729,22 @@ public void setLoggingEnabled(boolean loggingEnabled) {
         this.loggingEnabled = loggingEnabled;
     }
 
+
+    DeliveryParams getErrorApiDeliveryParams() {
```
[the remark this answers is not in the document]
Comment (fractalwrench) [reply bugsnag-android-core/src/main/java/com/bugsnag/android/Configuration.java]: This method is package-visible and therefore not accessible to the end-user, so Javadoc/main docs aren't required.

File: bugsnag-android-core/src/main/java/com/bugsnag/android/DefaultDelivery.kt [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+import java.net.HttpURLConnection
+import java.net.URL
+import java.nio.charset.Charset
+
+internal class DefaultDelivery(private val connectivity: Connectivity?) : Delivery {
+
+    override fun deliver(payload: SessionTrackingPayload,
+                         deliveryParams: DeliveryParams): DeliveryStatus {
+        val status = deliver(deliveryParams.endpoint, payload, deliveryParams.headers)
+        Logger.warn("Session API request finished with status $status")
```
[the remark this answers is not in the document]
Comment (fractalwrench) [reply bugsnag-android-core/src/main/java/com/bugsnag/android/DefaultDelivery.kt]: Changed this to `Logger.info` as the implementation logs the stacktrace/error message if something bad happens now.

**Comment by fractalwrench** [comment]
Discussed in-person and we have decided that the `Delivery` interface should remain the same, and logging should be improved in the `DefaultDelivery` class to help users diagnose error conditions.

**Comment by fractalwrench** [comment]
@bengourley I have updated this PR to improve the logging in the `DefaultDelivery` class. A successful request will now output the following information to Logcat:


>2019-08-02 13:59:35.227 25175-25212/com.bugsnag.android.example I/Bugsnag: Request completed with code 200, message: OK, headers: {null=[HTTP/1.1 200 OK], Access-Control-Allow-Origin=[*], Alt-Svc=[clear], Bugsnag-Event-Id=[123456789], Content-Length=[2], Content-Type=[text/plain; charset=utf-8], Date=[Fri, 02 Aug 2019 12:59:35 GMT], Via=[1.1 google], X-Android-Received-Millis=[1564750775226], X-Android-Response-Source=[NETWORK 200], X-Android-Selected-Protocol=[http/1.1], X-Android-Sent-Millis=[1564750775046]}
2019-08-02 13:59:35.228 25175-25212/com.bugsnag.android.example I/Bugsnag: Error API request finished with status DELIVERED

And an unsuccessful request would also contain the error body:

>2019-08-02 14:03:05.650 25317-25345/com.bugsnag.android.example I/Bugsnag: Request completed with code 415, message: Unsupported Media Type, headers: {null=[HTTP/1.1 415 Unsupported Media Type], Access-Control-Allow-Origin=[*], Alt-Svc=[clear], Content-Length=[165], Content-Type=[application/json], Date=[Fri, 02 Aug 2019 13:03:06 GMT], Via=[1.1 google], X-Android-Received-Millis=[1564750985636], X-Android-Response-Source=[NETWORK 415], X-Android-Selected-Protocol=[http/1.1], X-Android-Sent-Millis=[1564750985460]}
2019-08-02 14:03:05.656 25317-25345/com.bugsnag.android.example W/Bugsnag: Request error details: {"status":"error","errors":["Unsupported content type. Please provide 'application/json''. For documentation, see http://docs.bugsnagsessiontrackingapi.apiary.io/"]}
2019-08-02 14:03:05.656 25317-25345/com.bugsnag.android.example I/Bugsnag: Session API request finished with status FAILURE

Logging out the event ID was a great suggestion, as that should help us track down individual events for when people write into support 👍

**Pull request merged**